test/commit-and-push.ts

```diff
@@ -0,0 +1,426 @@
+      owner: origin.owner,
+      repo: origin.repo,
+      sha,
+      ref: 'heads/test-branch-name',
+    });
+    // restore
+    stub.restore();
+  });
+});
+
+describe('Commit and push function', () => {
+  const oldHeadSha = 'OLD-head-Sha-asdf1234';
+  const changes: Changes = new Map();
+  const origin: RepoDomain = {
+    owner: 'Foo',
+    repo: 'Bar',
+  };
+  const branchName = 'test-branch-name';
+  const message = 'Hello world';
+  describe('Works when everything works', () => {
+    it('Calls functions with correct parameter values', async () => {
+      const commitResponseData = await import(
+        './fixtures/get-commit-response.json'
+      );
+      const createTreeResponseData = await import(
+        './fixtures/create-tree-response.json'
+      );
+      const getCommitResponse = {
+        headers: {},
+        status: 200,
+        url: 'http://fake-url.com',
+        data: commitResponseData,
+      };
+      const createTreeResponse = {
+        headers: {},
+        status: 200,
+        url: 'http://fake-url.com',
+        data: createTreeResponseData,
+      };
+
+      const createCommitResponseData = await import(
+        './fixtures/create-commit-response.json'
+      );
+      const createCommitResponse = {
+        headers: {},
+        status: 201,
+        url: 'http://fake-url.com',
+        data: createCommitResponseData,
+      };
+      // setup
+      const stubGetCommit = sinon
+        .stub(octokit.git, 'getCommit')
+        .resolves(getCommitResponse);
+      const stubCreateTree = sinon
+        .stub(octokit.git, 'createTree')
+        .resolves(createTreeResponse);
+      const stubCreateCommit = sinon
+        .stub(octokit.git, 'createCommit')
+        .resolves(createCommitResponse);
+      const stubUpdateRef = sinon.stub(octokit.git, 'updateRef');
+      // tests
+      await handler.commitAndPush(
+        logger,
+        octokit,
+        oldHeadSha,
+        changes,
+        origin,
+        branchName,
+        message
+      );
+      sinon.assert.calledOnceWithExactly(stubGetCommit, {
+        owner: origin.owner,
+        repo: origin.repo,
+        commit_sha: oldHeadSha,
+      });
+      sinon.assert.calledWithExactly(stubCreateTree, {
+        owner: origin.owner,
+        repo: origin.repo,
+        tree: [],
+        base_tree: getCommitResponse.data.tree.sha,
+      });
+      sinon.assert.calledOnceWithExactly(stubCreateCommit, {
+        owner: origin.owner,
+        repo: origin.repo,
+        message,
+        tree: createTreeResponse.data.sha,
+        parents: [oldHeadSha],
+      });
+      sinon.assert.calledOnceWithExactly(stubUpdateRef, {
+        owner: origin.owner,
+        repo: origin.repo,
+        sha: createCommitResponse.data.sha,
+        ref: 'heads/test-branch-name',
+      });
+      // restore
+      stubGetCommit.restore();
+      stubCreateTree.restore();
+      stubCreateCommit.restore();
+      stubUpdateRef.restore();
```

Comment:
Not entirely sure how sinon stubs work - do stubs persist across requests? If a test fails, will the stub be stuck for the rest of the tests?

Reply:
A common pattern is to collect a list of stubs to restore and in the teardown block, iterate over all of those registered stubs to clear them out. Alternatively, their might be a sinon method for clearing all stubs?

Reply:
According to [best practices](https://semaphoreci.com/community/tutorials/best-practices-for-spies-stubs-and-mocks-in-sinon-js) and some other sites, I have seen wrapping tests with `sinon.test()` and using `sandboxes`, but neither seem to work on my environment so I am using your former suggestion. I've pushed another commit using global stubs and restoring them before each test. It's more manual/explicit since I understand loops are frowned upon in tests.

Reply:
i.e. the following sandbox implementation did not work
```
const sandbox = sinon.createSandbox();
describe('test', () => {
  afterEach(() => {
    sandbox.restore();
  });
  it('tests octokit', async() => {
    sandbox.stub(octokit.git, 'createCommit').resolves(createCommitResponse);
  });
}
```
fails if I repeat this test suite.

Reply:
I toyed around with it more and got the tests to reset after each test run. The docs are outdated (the method signatures didn't work as in the examples)